File: Backend/src/test/java/lan/dk/podcastserver/manager/selector/ExtractorSelectorTest.java
package lan.dk.podcastserver.manager.selector;

import io.vavr.collection.HashSet;
import lan.dk.podcastserver.manager.worker.Extractor;
import lan.dk.podcastserver.manager.worker.noop.PassThroughExtractor;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.stubbing.Answer;
import org.springframework.context.ApplicationContext;
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;

import java.util.Set;

import static lan.dk.podcastserver.manager.selector.ExtractorSelector.NO_OP_EXTRACTOR;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.when;

/**
 * Created by kevin on 03/12/2017
 */
@RunWith(SpringJUnit4ClassRunner.class)
public class ExtractorSelectorTest {

    private ExtractorSelector selector;

    private @Mock PassThroughExtractor passThroughExtractor;
    private @Mock ApplicationContext applicationContext;
    private Set<Extractor> extractors;

    @Before
    @SuppressWarnings("unchecked")
    public void setUp() throws Exception {
        when(passThroughExtractor.compatibility(anyString())).thenCallRealMethod();
        when(applicationContext.getBean(any(Class.class))).then(findBean());

        extractors = HashSet.<Extractor>of(passThroughExtractor).toJavaSet();
        selector = new ExtractorSelector(applicationContext, extractors);
    }

    private Answer<Extractor> findBean() {
        return i -> extractors
                .stream()
                .filter(d -> i.getArgumentAt(0, Class.class) == d.getClass())
                .findFirst()
                .orElse(NO_OP_EXTRACTOR);
    }

    @Test
    public void should_return_no_op_if_empty_string() {
        /* When  */ Extractor extractorClass = selector.of("");
        /* Then  */ assertThat(extractorClass).isEqualTo(ExtractorSelector.NO_OP_EXTRACTOR);
    }

    @Test
    public void should_do_pass_through_for_now() {
        /* When  */ Extractor extractorClass = selector.of("http://www.podtrac.com/pts/redirect.mp3/twit.cachefly.net/audio/tnt/tnt1217/tnt1217.mp3");
        /* Then  */ assertThat(extractorClass).isEqualTo(passThroughExtractor);
    }

}

File: backend/src/test/kotlin/com/github/davinkevin/podcastserver/manager/selector/ExtractorSelectorTest.kt
package com.github.davinkevin.podcastserver.manager.selector

import com.github.davinkevin.podcastserver.manager.worker.noop.PassThroughExtractor
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.whenever
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.mockito.junit.jupiter.MockitoSettings
import org.mockito.quality.Strictness
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.context.ApplicationContext
import org.springframework.test.context.junit.jupiter.SpringExtension
import java.net.URI
import java.util.stream.Stream
import kotlin.reflect.KClass

/**
 * Created by kevin on 03/12/2017
 */
@ExtendWith(SpringExtension::class)
@MockitoSettings(strictness = Strictness.LENIENT)
class ExtractorSelectorTest(
    @Autowired val applicationContext: ApplicationContext
) {

    @MockBean private lateinit var passThroughExtractor: PassThroughExtractor

    private lateinit var extractor: ExtractorSelector

    @BeforeEach
    fun beforeEach() {
        val extractors = setOf(passThroughExtractor)

        extractors.forEach { whenever(it.compatibility(any())).thenCallRealMethod() }

        extractor = ExtractorSelector(applicationContext, extractors)
    }

    @Test
    fun `should return no op if empty string`() {
        /* When  */
        val extractorClass = extractor.of(URI("https://foo.bar.com"))
        /* Then  */
        assertThat(extractorClass).isInstanceOf(PassThroughExtractor::class.java)
    }

    @MethodSource("urlToExtractor")
    @DisplayName("should return")
    @ParameterizedTest(name = "{1} finder for {0}")
    fun `should return matching updater`(url: URI, type: KClass<*>) {
        /* When */
        val extractorClass = extractor.of(url)
        /* Then */
        assertThat(extractorClass).isInstanceOf(type.java)
    }

    companion object {
        @JvmStatic
        fun urlToExtractor() =
                Stream.of(
                        Arguments.of(URI("http://www.beinsports.com/france/replay/lexpresso"), PassThroughExtractor::class),
                        Arguments.of(URI("http://www.dailymotion.com/foo/bar"), PassThroughExtractor::class),
                        Arguments.of(URI("http://www.jeuxvideo.com/chroniques-video.htm"), PassThroughExtractor::class),
                        Arguments.of(URI("http://foo.bar.com/to/rss/file.xml"), PassThroughExtractor::class),
                        Arguments.of(URI("http://www.youtube.com/channel/UC_ioajefokjFAOI"), PassThroughExtractor::class)
                )
    }

}
